topology/nhlt/intel/ssp/ssp-process.c

```diff
@@ -0,0 +1,707 @@
+		break;
+	case SSP_FMT_CBC_CFP:
+		ssp->ssp_blob[di][hwi].ssc1 |= SSCR1_SCFR | SSCR1_SFRMDIR;
+		/* FIXME: this mode has not been tested */
+		break;
+	default:
+		fprintf(stderr, "ssp_calculate(): format & PROVIDER_MASK EINVAL\n");
+		return -EINVAL;
+	}
+
+	/* clock signal polarity */
+	switch (ssp->ssp_prm.hw_cfg[hwi].format & SSP_FMT_INV_MASK) {
+	case SSP_FMT_NB_NF:
+		break;
+	case SSP_FMT_NB_IF:
+		inverted_frame = 1; /* handled later with format */
+		break;
+	case SSP_FMT_IB_IF:
+		inverted_bclk = 1; /* handled later with bclk idle */
+		inverted_frame = 1; /* handled later with format */
+		break;
+	case SSP_FMT_IB_NF:
+		inverted_bclk = 1; /* handled later with bclk idle */
+		break;
+	default:
+		fprintf(stderr, "ssp_calculate: format & INV_MASK EINVAL\n");
+		return -EINVAL;
+	}
+
+	/* supporting bclk idle state */
+	if (ssp->ssp_prm.clks_control &
+		SSP_INTEL_CLKCTRL_BCLK_IDLE_HIGH) {
+		/* bclk idle state high */
+		ssp->ssp_blob[di][hwi].sspsp |= SSPSP_SCMODE((inverted_bclk ^ 0x3) & 0x3);
+	} else {
+		/* bclk idle state low */
+		ssp->ssp_blob[di][hwi].sspsp |= SSPSP_SCMODE(inverted_bclk);
+	}
+
+	/* ACS removed for now */
+	ssp->ssp_blob[di][hwi].ssc0 |= SSCR0_MOD;
+
+	/* Additional hardware settings */
+
+	/* Receiver Time-out Interrupt Disabled/Enabled */
+	ssp->ssp_blob[di][hwi].ssc1 |= (ssp->ssp_prm.quirks & SSP_INTEL_QUIRK_TINTE) ?
+		SSCR1_TINTE : 0;
+
+	/* Peripheral Trailing Byte Interrupts Disable/Enable */
+	ssp->ssp_blob[di][hwi].ssc1 |= (ssp->ssp_prm.quirks & SSP_INTEL_QUIRK_PINTE) ?
+		SSCR1_PINTE : 0;
+
+	/* Enable/disable internal loopback. Output of transmit serial
+	 * shifter connected to input of receive serial shifter, internally.
+	 */
+	ssp->ssp_blob[di][hwi].ssc1 |= (ssp->ssp_prm.quirks & SSP_INTEL_QUIRK_LBM) ?
+		SSCR1_LBM : 0;
+
+	/* Transmit data are driven at the same/opposite clock edge specified
+	 * in SSPSP.SCMODE[1:0]
+	 */
+	ssp->ssp_blob[di][hwi].ssc2 |= (ssp->ssp_prm.quirks & SSP_INTEL_QUIRK_SMTATF) ?
+		SSCR2_SMTATF : 0;
+
+	/* Receive data are sampled at the same/opposite clock edge specified
+	 * in SSPSP.SCMODE[1:0]
+	 */
+	ssp->ssp_blob[di][hwi].ssc2 |= (ssp->ssp_prm.quirks & SSP_INTEL_QUIRK_MMRATF) ?
+		SSCR2_MMRATF : 0;
+
+	/* Enable/disable the fix for PSP consumer mode TXD wait for frame
+	 * de-assertion before starting the second channel
+	 */
+	ssp->ssp_blob[di][hwi].ssc2 |= (ssp->ssp_prm.quirks & SSP_INTEL_QUIRK_PSPSTWFDFD) ?
+		SSCR2_PSPSTWFDFD : 0;
+
+	/* Enable/disable the fix for PSP provider mode FSRT with dummy stop &
+	 * frame end padding capability
+	 */
+	ssp->ssp_blob[di][hwi].ssc2 |= (ssp->ssp_prm.quirks & SSP_INTEL_QUIRK_PSPSRWFDFD) ?
+		SSCR2_PSPSRWFDFD : 0;
+
+#if 0
+	if (!ssp->ssp_prm.hw_cfg[hwi].mclk_rate ||
+	    ssp->ssp_prm.hw_cfg[hwi].mclk_rate > ssp_freq[MAX_SSP_FREQ_INDEX].freq) {
+		fprintf("ssp_calculate(): invalid MCLK = %d Hz (valid < %d)\n",
+			ssp->ssp_prm.hw_cfg[hwi].mclk_rate,
+			ssp_freq[MAX_SSP_FREQ_INDEX].freq);
+		return -EINVAL;
+	}
+#endif
```

Comment:
??

Reply:
hmm this is a check against maximum possible ssp frequency, which I understand is platform dependent. So not sure I can get this somewhere nicely.

Reply:
for e.g. in tgl this would be 96000000, so if this is added to the topology I can do the check

Reply:
but yeah will now just check for 0 mclk